psiminer-core/src/main/kotlin/psi/graphs/edgeProviders/php/PhpControlFlowEdgeProvider.kt

```diff
@@ -0,0 +1,160 @@
+     * For some strange reason in PhpInstruction we can only access its predecessors.
+     * This function uses this information to create set of successors for each instruction.
+     */
+    private fun calculateSuccessors(instructions: List<PhpInstruction>): Array<out Set<Int>> {
+        val n = instructions.size
+        val successors = Array<MutableSet<Int>>(n) { mutableSetOf() }
+        instructions.forEach { instruction ->
+            instruction.predecessors.forEach { predecessor ->
+                successors[predecessor.num()].add(instruction.num())
+            }
+        }
+        return successors
+    }
+
+    private fun computeNextNonBranchingInstructions(
+        instructions: List<PhpInstruction>,
+        elements: List<PsiElement?>,
+        successors: Array<out Set<Int>>
+    ): Array<out Set<Int>> {
+        val n = instructions.size
+        val nextInstructions = Array<MutableSet<Int>>(n) { mutableSetOf() }
+        val used = BooleanArray(n) { false }
+
+        fun dfsFromInstruction(index: Int) {
+            if (used[index]) {
+                return
+            }
+            used[index] = true
+            for (toIndex in successors[index]) {
+                val toElement = elements[toIndex]
+                if (!toElement.correspondsToBranchingInstruction()) {
+                    nextInstructions[index].add(toIndex)
+                } else {
+                    dfsFromInstruction(toIndex)
+                    nextInstructions[index].addAll(nextInstructions[toIndex])
+                }
+            }
+        }
+
+        for (index in 0 until n) {
+            dfsFromInstruction(index)
+        }
+        return nextInstructions
+    }
+
+    private fun PsiElement?.correspondsToBranchingInstruction(): Boolean =
+        this is ControlStatement || this is PhpReturn
+
+    private fun provideEdgesForInstruction(
+        index: Int,
+        instructions: List<PhpInstruction>,
+        elements: List<PsiElement?>,
+        successors: Array<out Set<Int>>,
+        nextInstructions: Array<out Set<Int>>,
+    ): MutableList<Edge> {
+        val newEdges = mutableListOf<Edge>()
+        val element = elements[index] ?: return mutableListOf()
+        if (!element.correspondsToBranchingInstruction()) {
+            nextInstructions[index].forEach { toIndex ->
+                val toElement = elements[toIndex]
+                if (toElement != element && toElement != null) {
+                    newEdges.add(Edge(element, toElement, EdgeType.ControlFlow))
+                }
+            }
+        }
+        for (toIndex in successors[index]) {
+            val edgeType = EdgeType.ControlElement
+            val toElement = elements[toIndex] ?: continue
+            if (
+                element != toElement &&
+                (element.correspondsToBranchingInstruction() || toElement.correspondsToBranchingInstruction())
+            ) {
+                newEdges.add(Edge(element, toElement, edgeType))
```

Comment:
You can remove the edgeType variable here and substitute the actual type. It will be a bit more readable.